src/code/z_demo.c

```diff
@@ -2220,6 +2244,14 @@ void CutsceneHandler_RunScript(PlayState* play, CutsceneContext* csCtx) {
         }
     }
 }
+#else
+void CutsceneHandler_RunScript(PlayState* play, CutsceneContext* csCtx) {
+    if (gSaveContext.save.cutsceneIndex >= 0xFFF0) {
+        csCtx->curFrame++;
+        Cutscene_ProcessScript(play, csCtx, play->csCtx.script);
+    }
+}
+#endif
```

Comment:
I made an entirely new copy of this function for retail, cause ifdeffing inside the function got messy really fast. If anyone has other ideas tho lmk

Reply:
Personally I think it would be better to put the ifdef inside the function, but at the very top
```c
void CutsceneHandler_RunScript(PlayState* play, CutsceneContext* csCtx) {
#ifdef OOT_DEBUG
    // ...
#else
    // ...
#endif
}
```
This is nicer for the IDE in stuff like doing `ctrl-click` on the name of the function so it can lead you to the source, otherwise it would show you a list of possible functions, which is kinda annoying

Reply:
Possibly.
An alternative to that I guess is two separate ifdefs, one for the debug print info, and another for the processscript. Still messy and bad imo.
More opinions needed
```c
void CutsceneHandler_RunScript(PlayState* play, CutsceneContext* csCtx) {
#ifdef OOT_DEBUG
    if (0) {} // Necessary to match

    if (gSaveContext.save.cutsceneIndex >= 0xFFF0) {
        if (0) {} // Also necessary to match

        if (BREG(0) != 0) {
            Gfx* displayList;
            Gfx* prevDisplayList;

            OPEN_DISPS(play->state.gfxCtx, "../z_demo.c", 4101);

            prevDisplayList = POLY_OPA_DISP;
            displayList = Graph_GfxPlusOne(POLY_OPA_DISP);
            gSPDisplayList(OVERLAY_DISP++, displayList);
            Cutscene_DrawDebugInfo(play, &displayList, csCtx);
            gSPEndDisplayList(displayList++);
            Graph_BranchDlist(prevDisplayList, displayList);
            POLY_OPA_DISP = displayList;

            CLOSE_DISPS(play->state.gfxCtx, "../z_demo.c", 4108);
        }
#endif
        csCtx->curFrame++;

#ifdef OOT_DEBUG
        if (R_USE_DEBUG_CUTSCENE) {
            Cutscene_ProcessScript(play, csCtx, gDebugCutsceneScript);
        } else {
            Cutscene_ProcessScript(play, csCtx, play->csCtx.script);
        }
#else
        Cutscene_ProcessScript(play, csCtx, play->csCtx.script);
#endif
    }
}
```

Reply:
also the required if 0's hint more toward cadmics idea of `if (OOT_DEBUG)`

Reply:
> also the required if 0's hint more toward cadmics idea of `if (OOT_DEBUG)`

Could you match this function with that idea and removing the if0s

Reply:
yes

of note, just using `OOT_DEBUG_DEF` in the condition at the end is enough to remove the if(0) required in debug. I thought it wouldnt, because typically this means more { } blocks are needed. but seemingly not in this case? or I just dont understand ido's optimiation, which is likely 
```c
#define OOT_DEBUG_DEF 0

void CutsceneHandler_RunScript(PlayState* play, CutsceneContext* csCtx) {
    if (gSaveContext.save.cutsceneIndex >= 0xFFF0) {
        if (OOT_DEBUG_DEF) {
            if (BREG(0) != 0) {
                Gfx* displayList;
                Gfx* prevDisplayList;

                OPEN_DISPS(play->state.gfxCtx, "../z_demo.c", 4101);

                prevDisplayList = POLY_OPA_DISP;
                displayList = Graph_GfxPlusOne(POLY_OPA_DISP);
                gSPDisplayList(OVERLAY_DISP++, displayList);
                Cutscene_DrawDebugInfo(play, &displayList, csCtx);
                gSPEndDisplayList(displayList++);
                Graph_BranchDlist(prevDisplayList, displayList);
                POLY_OPA_DISP = displayList;

                CLOSE_DISPS(play->state.gfxCtx, "../z_demo.c", 4108);
            }
        }
        csCtx->curFrame++;

        if (OOT_DEBUG_DEF && R_USE_DEBUG_CUTSCENE) {
            Cutscene_ProcessScript(play, csCtx, gDebugCutsceneScript);
        } else {
            Cutscene_ProcessScript(play, csCtx, play->csCtx.script);
        }
    }
}
```

Reply:
you can actually move the debug define into the first conditional with && too

Reply:
I wasnt too convinced with this theory yet, but this is a pretty huge example. any other way of writing this with ifdefs just sucks :D

Reply:
Yeah
My main worry now is warnings spam about unreachable code
Idk about ido but I imagine gcc would complain

Reply:
Gcc *currently* does not warn for unreachable code. It used to be a feature, it was removed 10+ years ago because it was bugged
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=46476

But ideally a compiler should warn about dead code. e.g. if someone ever gets clang working (and its mips support gets less bad), clang does currently warn about dead code

https://stackoverflow.com/questions/45598080/clang-gcc-disable-unused-code-warning-as-a-result-of-conditional-macro
a comment says (some versions of?) clang are smart enough to not warn about unreachable code in these cases (i.e. recognize the macro may expand to not-0)

Note this post also links to
https://stackoverflow.com/questions/1644868/define-macro-for-debug-printing-in-c
which strongly advocates for `if(debug)` over `#ifdef debug`

But overall I'm not finding much on the internet about this which leads me to believe the `#ifdef` way is how traditional C handles this (it's what I've typically encountered personally, whereas I've seen C++ work with constexpr in these cases)


So currently my take would be "use #ifdef unless if() is really really worth it" (like in this case where it avoids code duplication, and if0s as a bonus)